locopy/redshift.py

```diff
@@ -0,0 +1,541 @@
+    """Locopy class which manages connections to Redshift.  Inherits ``Base`` and implements the
+    specific ``COPY`` and ``UNLOAD`` functionality.
+
+    If any of host, port, dbname, user and password are not provided, a config_yaml file must be
+    provided with those parameters in it. Please note ssl is always enforced when connecting.
+
+    Parameters
+    ----------
+    profile : str, optional
+        The name of the AWS profile to use which is typical stored in the
+        ``credentials`` file.  You can also set environment variable
+        ``AWS_DEFAULT_PROFILE`` which would be used instead.
+
+    kms_key : str, optional
+        The KMS key to use for encryption
+        If kms_key Defaults to ``None`` then the AES256 ServerSideEncryption
+        will be used.
+
+    dbapi : DBAPI 2 module, optional
+        A PostgreSQL database adapter which is Python DB API 2.0 compliant
+        (``psycopg2``, ``pg8000``, etc.)
+
+    host : str, optional
+        Host name of the Redshift cluster to connect to.
+
+    port : int, optional
+        Port which connection will be made to Redshift.
+
+    dbname : str, optional
+        Redshift database name.
+
+    user : str, optional
+        Redshift users username.
+
+    password : str, optional
+        Redshift users password.
+
+    config_yaml : str, optional
+        String representing the file location of the credentials.
+
+    **kwargs
+        Optional keyword arguments.
+
+    Attributes
+    ----------
+    profile : str
+        String representing the AWS profile for authentication
+
+    kms_key : str
+        String representing the s3 kms key
+
+    session : boto3.Session
+        Hold the AWS session credentials / info
+
+    s3 : botocore.client.S3
+        Hold the S3 client object which is used to upload/delete files to S3
+
+    dbapi : DBAPI 2 module
+        database adapter which is Python DBAPI 2.0 compliant
+
+    host : str
+        Database host name
+
+    port : int
+        Database port number
+
+    database : str
+        Database name.
+
+    user : str
+        Database username.
+
+    password : str
+        Database password.
+
+    extra_conn : dict
+        Dictionary of additonal connection items
+
+    conn : dbapi.connection
+        DBAPI connection instance
+
+    cursor : dbapi.cursor
+        DBAPI cursor instance
+
+    Raises
+    ------
+    CredentialsError
+        Database credentials are not provided or valid
+
+    S3Error
+        Error initializing AWS Session (ex: invalid profile)
+
+    S3CredentialsError
+        Issue with AWS credentials
+
+    S3InitializationError
+        Issue initializing S3 session
+    """
+
+    def __init__(
+        self,
+        profile=None,
+        kms_key=None,
+        dbapi=None,
+        host=None,
+        port=None,
+        database=None,
+        user=None,
+        password=None,
+        config_yaml=None,
+        **kwargs
+    ):
+        S3.__init__(self, profile, kms_key)
```

Comment:
I think this inheritance and `__init__` pattern would mean that you are forced to have the ability to set up S3 sessions to use any of Redshift functionality (like just running arbitrary SQL and pulling out through the cursor), vs. currently where you can still use `Cmd` without worrying about your tokens 😁

Is there a way to get the best of both worlds, where either:
- `S3.__init__` is only called when some functionality from S3 is needed (Not sure if that's possible), or
- `S3.__init__` wouldn't actually set up a session unless something was called that needed the session?  Like maybe add a `@needs_connection` decorator to `S3` functions that need a session, which would call `self._set_session` if `self` didn't have a session set already?

Reply:
Maybe we rename it so `Base` --> `Redshift` and `Redshift` --> `Locopy`?
That would solve this issue here. If you just want to run arbitrary SQL code `Base` should be enough. While it is nice to provide the functionality just using `psycopg2` would be a lower entry to run just SQL IMO. `locopy` is meant to solve the specific problem of loading/unloading data.